core/services/pipeline/graph.go

```diff
@@ -123,55 +124,78 @@ func (g TaskDAG) outputs() []*taskDAGNode {
 	return outputs
 }
 
-type taskDAGNode struct {
+type TaskDAGNode struct {
 	graph.Node
 	g     *TaskDAG
 	dotID string
 	attrs map[string]string
 }
 
-func (n *taskDAGNode) DOTID() string {
+func NewTaskDAGNode(n graph.Node, dotID string, attrs map[string]string) *TaskDAGNode {
+	return &TaskDAGNode{
+		Node:  n,
+		attrs: attrs,
+		dotID: dotID,
+	}
+}
+
+func (n *TaskDAGNode) SetDAG(g *TaskDAG) {
+	n.g = g
+}
+
+func (n *TaskDAGNode) DOTID() string {
 	return n.dotID
 }
 
-func (n *taskDAGNode) SetDOTID(id string) {
+func (n *TaskDAGNode) SetDOTID(id string) {
 	n.dotID = id
 }
 
-func (n *taskDAGNode) String() string {
+func (n *TaskDAGNode) String() string {
 	return n.dotID
 }
 
-func (n *taskDAGNode) SetAttribute(attr encoding.Attribute) error {
+func (n *TaskDAGNode) SetAttribute(attr encoding.Attribute) error {
 	if n.attrs == nil {
 		n.attrs = make(map[string]string)
 	}
 	n.attrs[attr.Key] = attr.Value
 	return nil
 }
 
-func (n *taskDAGNode) inputs() []*taskDAGNode {
-	var nodes []*taskDAGNode
+func (n *TaskDAGNode) Attributes() []encoding.Attribute {
+	var r []encoding.Attribute
+	for k, v := range n.attrs {
+		r = append(r, encoding.Attribute{Key: k, Value: v})
+	}
+	sort.Slice(r, func(i, j int) bool {
```

Comment:
sorting just to ensure deterministic ordering of node attributes, helps with the expected test dags

Reply:
Makes sense, let's add this as a comment tho